src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,351 @@
+            .ConfigureAwait(false);
+        var responseJson = JsonSerializer.Serialize(result, handler.ResponseType);
+        return responseJson;
+    }
+
+    public async ValueTask<string> HandleExtensionDocumentMessageAsync(
+        Document document,
+        string messageName,
+        string jsonMessage,
+        CancellationToken cancellationToken)
+    {
+        IExtensionDocumentMessageHandlerWrapper handler;
+        lock (_lockObject)
+        {
+            if (!_documentHandlers.TryGetValue(messageName, out handler!))
+            {
+                throw new InvalidOperationException($"No document handler found for message {messageName}.");
+            }
+        }
+
+        // Any exception thrown in this method is left to bubble up to the extension.
+        var message = JsonSerializer.Deserialize(jsonMessage, handler.MessageType);
+        var result = await handler.ExecuteAsync(message, document, cancellationToken)
+            .ConfigureAwait(false);
+        var responseJson = JsonSerializer.Serialize(result, handler.ResponseType);
+        return responseJson;
+    }
+
+    public ValueTask UnregisterExtensionAsync(
+        string assemblyFilePath,
+        CancellationToken cancellationToken)
+    {
+        var assemblyFileName = Path.GetFileName(assemblyFilePath);
+        var assemblyFolderPath = Path.GetDirectoryName(assemblyFilePath)
+            ?? throw new InvalidOperationException($"Unable to get the directory name for {assemblyFilePath}.");
+
+        try
+        {
+            CustomMessageHandlerExtension? extension = null;
+            lock (_lockObject)
+            {
+                if (_extensions.TryGetValue(assemblyFolderPath, out extension))
+                {
+                    if (extension.Assemblies.TryGetValue(assemblyFileName, out var extensionAssembly))
+                    {
+                        extension.Assemblies.Remove(assemblyFileName);
+                        UnregisterHandlersForAssembly(extensionAssembly);
+                    }
+
+                    if (extension.Assemblies.Count == 0)
+                    {
+                        _extensions.Remove(assemblyFolderPath);
+                        foreach (var assembly in extension.Assemblies.Values)
+                        {
+                            UnregisterHandlersForAssembly(assembly);
+                        }
+                    }
+                }
+            }
+
+            extension?.AnalyzerAssemblyLoader.Dispose();
+        }
+        catch (Exception e) when (LogAndPropagate(e))
+        {
+            // unreachable
+        }
+
+        return ValueTask.CompletedTask;
+
+        bool LogAndPropagate(Exception e)
+        {
+            Logger.Log(
+                FunctionId.CustomMessageHandlerService_UnloadCustomMessageHandlerAsync,
+                $"Error unregistering {assemblyFilePath}: {e}",
+                LogLevel.Error);
+            return false;
+        }
```

Comment:
would love if we were using FatalError here, so taht we get actual telemetry on these things happening.